splendor/src/state.rs

```diff
@@ -0,0 +1,361 @@
+            dev_level3: Some(self.board_state.dev_cards[2].get_pub_state()),
+            resources_on_board: Some(self.board_state.resources_on_board.clone()),
+            nobles_on_board: self.board_state.nobles_on_board.clone(),
+            player_states
+        }
+    }
+}
+
+pub struct BoardState {
+    pub dev_cards: [DevCardsLevelState; 3],
+    pub resources_on_board: ResourceMap,
+    pub nobles_on_board: Vec<Noble>
+}
+
+impl BoardState {
+    pub fn new() -> BoardState {
+        BoardState {
+            dev_cards: [
+                DevCardsLevelState::new_level1(),
+                DevCardsLevelState::new_level2(),
+                DevCardsLevelState::new_level3(),
+            ],
+            resources_on_board: ResourceMap {
+                entries: vec![
+                    ResourceEntry { resource_type: ResourceType::Agate as i32, number: 7 },
+                    ResourceEntry { resource_type: ResourceType::Emerald as i32, number: 7 },
+                    ResourceEntry { resource_type: ResourceType::Diamond as i32, number: 7 },
+                    ResourceEntry { resource_type: ResourceType::Ruby as i32, number: 7 },
+                    ResourceEntry { resource_type: ResourceType::Sapphire as i32, number: 7 },
+                    ResourceEntry { resource_type: ResourceType::Gold as i32, number: 5 },
+                ]
+            },
+            nobles_on_board: BoardState::rand_nobles()
+        }
+    }
+
+    pub fn rand_nobles() -> Vec<Noble> {
+        let mut nobles = vec![];
+        for i in 0..5 {
+            loop {
+                // 3 nobles like 3, 3, 3 and 2 nobles like 4, 4
+                let resource_type_num = if i < 3 { 3 } else { 2 };
+                let numbers = if i < 3 { vec![3, 3, 3] } else { vec![4, 4] };
+
+                let rands = get_multi_rand_nums(
+                    resource_type_num, ResourceType::Agate as i32, ResourceType::Sapphire as i32);
+                let noble = get_a_rand_noble(rands, numbers);
+                if nobles.iter().all(|n| *n != noble) {
+                    // no repatitive noble
+                    nobles.push(noble);
+                    break;
+                }
+            }
+        }
+        nobles
+    }
+
+    pub fn take_resource(&mut self, resource_type: i32) {
+        let entry = &mut self.resources_on_board.entries[resource_type as usize];
+        assert_eq!(entry.resource_type, resource_type);
+        entry.number -= 1;
+        assert!(entry.number >= 0);
+    }
+
+    pub fn return_resource(&mut self, res_map: ResourceMap) {
+        for entry in res_map.entries.iter() {
+            self.resources_on_board.entries[entry.resource_type as usize].number += entry.number;
+        }
+    }
+}
+
+pub struct DevCardsLevelState {
+    pub deck: Vec<Development>,
+    pub cards_on_board: Vec<Development>
+}
+
+impl DevCardsLevelState {
+    pub fn new_level1() -> DevCardsLevelState {
```

Comment:
these templates looks like some constant variables. Could we have a centralized definition?
and, do we really need a random algo? the cards are fixed?

Reply:
yep, it's better to put these templates somewhere together.
as for the random algo, the aim is to make the game more undeterministic. (because it's hard to balance each resource type if use fixed cards)